deltakit-core/src/deltakit_core/decoding_graphs/_decoding_graph.py

```diff
@@ -153,35 +167,35 @@ class DecodingHyperMultiGraph(HyperMultiGraph[Tuple[DecodingHyperEdge, int]]):
     def __init__(
         self,
         edge_data: Iterable[
-            Union[
-                DecodingHyperEdge, Tuple[int, ...], Tuple[DecodingHyperEdge, EdgeRecord]
-            ]
+            DecodingHyperEdge | tuple[int, ...] | tuple[DecodingHyperEdge, EdgeRecord]
         ],
-        detector_records: Optional[Dict[int, DetectorRecord]] = None,
+        detector_records: dict[int, DetectorRecord] | None = None,
     ):
-        edge_records: Dict[Tuple[DecodingHyperEdge, int], EdgeRecord] = {}
-        edges: List[Tuple[DecodingHyperEdge, int]] = []
-        edge_count: Dict[DecodingHyperEdge, int] = defaultdict(int)
+        edge_records: dict[tuple[DecodingHyperEdge, int], EdgeRecord] = {}
+        edges: list[tuple[DecodingHyperEdge, int]] = []
+        edge_count: dict[DecodingHyperEdge, int] = defaultdict(int)
 
         for data in edge_data:
             if isinstance(data, DecodingHyperEdge):
                 edge = data
                 edge_record = EdgeRecord()
-            elif isinstance(data, tuple) and isinstance(data[-1], EdgeRecord):
-                edge, edge_record = data  # type: ignore
-            elif isinstance(data, tuple) and isinstance(data[-1], int):
-                edge = DecodingHyperEdge(data)  # type: ignore
+            elif _is_decoding_hyper_edge_and_edge_record(data):
+                edge, edge_record = data
+            elif _is_tuple_of_ints(data):
+                edge = DecodingHyperEdge(data)
```

Comment:
To double-check, introduction of `TypeGuard`, see https://typing.python.org/en/latest/spec/narrowing.html